intel-sgx/dcap-artifact-retrieval/src/provisioning_client/pccs.rs

```diff
@@ -0,0 +1,670 @@
+        QeIdIn {
+            api_version: self.api_version,
+        }
+    }
+}
+
+/// Implementation of Get Intel's QE Identity API (section 3.4 of [reference]).
+///
+/// [reference]: <https://download.01.org/intel-sgx/sgx-dcap/1.22/linux/docs/SGX_DCAP_Caching_Service_Design_Guide.pdf>
+impl<'inp> ProvisioningServiceApi<'inp> for QeIdApi {
+    type Input = QeIdIn;
+    type Output = QeIdentitySigned;
+
+    fn build_request(&self, input: &Self::Input) -> Result<(String, Vec<(String, String)>), Error> {
+        let api_version = input.api_version as u8;
+        let url = format!(
+            "{}/sgx/certification/v{}/qe/identity",
+            self.base_url, api_version,
+        );
+        Ok((url, Vec::new()))
+    }
+
+    fn validate_response(&self, status_code: StatusCode) -> Result<(), Error> {
+        match &status_code {
+            StatusCode::Ok => Ok(()),
+            StatusCode::NotFound => Err(Error::PCSError(
+                status_code,
+                "QE identity information cannot be found",
+            )),
+            StatusCode::InternalServerError => Err(Error::PCSError(
+                status_code,
+                "PCCS suffered from an internal server error",
+            )),
+            StatusCode::BadGateway => Err(Error::PCSError(
+                status_code,
+                "Unable to retrieve the collateral from the Intel SGX PCS",
+            )),
+            _ => Err(Error::PCSError(
+                status_code,
+                "Unexpected response from PCCS server",
+            )),
+        }
+    }
+
+    fn parse_response(
+        &self,
+        response_body: String,
+        response_headers: Vec<(String, String)>,
+        _api_version: PcsVersion,
+    ) -> Result<Self::Output, Error> {
+        let ca_chain = parse_issuer_header(&response_headers, ENCLAVE_ID_ISSUER_CHAIN_HEADER)?;
+        let id = QeIdentitySigned::parse(&response_body, ca_chain)?;
+        Ok(id)
+    }
+}
+
+#[cfg(all(test, feature = "reqwest"))]
+mod tests {
+    use std::hash::{DefaultHasher, Hash, Hasher};
+    use std::path::PathBuf;
+    use std::time::Duration;
+
+    use pcs::PckID;
+
+    use super::Client;
+    use crate::provisioning_client::{
+        test_helpers, PccsProvisioningClientBuilder, PcsVersion, ProvisioningClient,
+    };
+    use crate::{reqwest_client_insecure_tls, ReqwestClient};
+
+    const PCKID_TEST_FILE: &str = "./tests/data/pckid_retrieval.csv";
+    const OUTPUT_TEST_DIR: &str = "./tests/data/";
+    const TIME_RETRY_TIMEOUT: Duration = Duration::from_secs(180);
+    const PCCS_URL: &'static str = "https://localhost:8081";
+
+    fn make_client(api_version: PcsVersion) -> Client<ReqwestClient> {
+        PccsProvisioningClientBuilder::new(api_version, PCCS_URL)
+            .set_retry_timeout(TIME_RETRY_TIMEOUT)
+            .build(reqwest_client_insecure_tls())
+    }
+
+    #[test]
+    #[ignore = "needs a running PCCS service"] // FIXME
```

Comment:
If want, this actually can be fixed by using https://docs.rs/httpmock/latest/httpmock/

Reply:
I want to test it against the real API to catch bugs in this code. We can enable this once we have a running caching service.